kernel: enable Multi-Path TCP for !SMALL_FLASH targets

dangowrt opened the pull request:
Expose Kernel's CONFIG_MPTCP option and enable it by default for !SMALL_FLASH targets.

dangowrt committed:
kernel: enable Multi-Path TCP for !SMALL_FLASH targets

Expose Kernel's CONFIG_MPTCP option and enable it by default for
!SMALL_FLASH targets.

The idea behind enabling it by default is to allow users of the binary
distribution to make use of MPTCP tunneling for link aggregation.

Using MPTCP for link aggregation is an often discussed topic in the
forum and there is even a whole OpenWrt fork (MPTCPRouter) just for that.

Enabling the kernel-side of the story by default will allow using MPTCP
on vanilla OpenWrt without having to build anything from source.

See also https://openwrt.org/docs/guide-user/network/mptcp

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

dangowrt commented:
@arinc9 Let me know if any additional patches are needed on top of that for MPTCP.

nbd168 commented:
Why enable it by default?

nbd168 commented:
Makes sense, thanks. Please put that into the commit description, so that other people can figure out the reason as well.

hauke commented on config/Config-kernel.in:
```diff
@@ -1204,6 +1204,18 @@ config KERNEL_PAGE_POOL_STATS
+
+config KERNEL_MPTCP_IPV6
+	bool "IPv6 support for Multipath TCP"
+	depends on KERNEL_MPTCP
```
Please make this option depend on `KERNEL_IPV6` or move it to the block inside the `if KERNEL_IPV6` block.

dangowrt committed:
kernel: add MultiPath TCP diag kernel module

Add INET diag socket monitoring interface module for MultiPath TCP.
It allows using native Linux socket diagnostic tools such as 'ss' for
Multipath TCP connections.

Co-authored-by: Rodrigo B. de Sousa Martins <rodrigo.sousa.577@gmail.com>
Signed-off-by: sKy King <29267720+sKyissKy@users.noreply.github.com>
Link: https://github.com/openwrt/openwrt/pull/12884
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
Signed-off-by: Daniel Golle <daniel@makrotopia.org>

dangowrt commented:
Thank you @hauke for bringing the existing pull requests to my attention. I've picked the module from #14323 and made `CONFIG_KERNEL_MPTCP_IPV6` dependent on `CONFIG_IPV6`.

Neustradamus commented:
@dangowrt: Nice PR, thanks!

RomanovDanila commented:
what about bpf schedulers from mptcp next? 

https://github.com/Ysurac/openmptcprouter/blob/develop/6.6/target/linux/generic/hack-6.6/999-mptcp-bpf.patch

https://github.com/Ysurac/openmptcprouter/blob/develop/common/include/bpf_mptcp.mk

mptcp-bpf-* from openmptcprouter-feeds
https://github.com/Ysurac/openmptcprouter-feeds

https://github.com/multipath-tcp/mptcp_net-next/commits/export/net/mptcp
https://github.com/multipath-tcp/mptcp_net-next/commits/export/tools

matttbe commented:
Thank you for looking at that!

> what about bpf schedulers from mptcp next?

If these BPF schedulers have not been upstreamed, that's because they are not ready. I would personally not recommend using them in production.

dangowrt commented:
> what about bpf schedulers from mptcp next?

I'm open to include those in a future PR backporting the necessary changes from net-next once they are available there.

Neustradamus commented:
@dangowrt: Good job!

systemcrash commented:
Hi @dangowrt - you linked to the wiki page in your commit description, which talks about options which I cannot find referenced anywhere. It's my understanding that procd and netifd don't yet know anything about e.g. `backup`, `master`, `handover` etc. Should those even be in there?

Are those for something older in another project?

matttbe commented:
> Hi @dangowrt - you linked to the wiki page in your commit description, which talks about options which I cannot find referenced anywhere. It's my understanding that procd and netifd don't yet know anything about e.g. `backup`, `master`, `handover` etc. Should those even be in there?
> 
> Are those for something older in another project?

Indeed, it looks like it is linked to another project (Open MPTCP Router?). I guess everything from the [Configuration](https://openwrt.org/docs/guide-user/network/mptcp#configuration) section should be removed, no?

Also, don't hesitate to reference [mptcp.dev](https://www.mptcp.dev) instead of [multipath.org](https://multipath-tcp.org)

matttbe commented:
Regarding the [mptcp](https://openwrt.org/docs/guide-user/network/mptcp) wiki page, I don't think it is clear what needs to be done for the configuration. Is the idea to install `mptcpd` so it can set the MPTCP endpoints automatically. Or is the goal to have `netifd` (or something else reading `/etc/config/network`) setting the MPTCP limits and endpoints when a new IP is available on certain interfaces?

In the meantime, I guess a script in `/etc/hotplug.d/iface/` could be used to set the endpoints manually with `ip mptcp` (requiring `ip-full` package).

Neustradamus commented:
@matttbe: Maybe you can connect on Wiki to update?

matttbe commented:
@Neustradamus I could ask the admins to create an account for me to change the website, but for the rest, I don't know how MPTCP is supposed to be configured on OpenWrt.

All I can see is the wiki page history: only the introduction ("Start" and "Multipath TCP") has been modified recently. The rest is confusing: it looks outdated and no longer valid. I could remove it, but I don't know what to replace it with: what's the recommended way to setup MPTCP on OpenWrt if this cannot be done via `uci`?

On mptcp.dev, I recommend using NM and mptcpd, but not sure if that's a good idea on OpenWrt: I guess NM is not doing to be used here, and mptcpd is currently not integrated with (l)uci. Also, for the manual configuration, `ip mptcp` is not available by default. For the workaround with `hotplug.d`, I don't have any setup to validate it (and currently no need to).

systemcrash commented:
I had a poke around and found the same thing. Some info there is only relevant to the fork and is quite confusing.
